packages/database/test/helpers/syncpoint-util.ts

```diff
@@ -0,0 +1,603 @@
+            if (expectedEvent.type !== 'value') {
+              if (expectedEvent.name !== actualEvent.snapshot.key) {
+                return false;
+              }
+              if (
+                expectedEvent.type !== 'child_removed' &&
+                expectedEvent.prevName !== actualEvent.prevName
+              ) {
+                return false;
+              }
+            }
+            // make sure the snapshots match
+            const snapHash = actualEvent.snapshot._node.hash();
+            const expectedHash = nodeFromJSON(expectedEvent.data).hash();
+            return snapHash === expectedHash;
+          } else {
+            return false;
+          }
+        });
+        if (!found) {
+          // console.log(actualEvent);
+          throw new Error('Could not find matching expected event');
+        }
+        currentExpected = currentExpected.slice(i);
+        currentActual = currentActual.slice(i);
+      }
+    };
+
+    const parseQuery = function (q, params) {
+      if (!('tag' in params)) {
+        throw new Error('Non-default queries must have a tag');
+      }
+      // eslint-disable-next-line guard-for-in
+      for (const paramName in params) {
+        const paramValue = params[paramName];
+        if (paramName === 'limitToFirst') {
+          q = query(q, limitToFirst(paramValue));
+        } else if (paramName === 'limitToLast') {
+          q = query(q, limitToLast(paramValue));
+        } else if (paramName === 'startAt') {
+          q = query(q, startAt(paramValue.index, paramValue.name));
+        } else if (paramName === 'endAt') {
+          q = query(q, endAt(paramValue.index, paramValue.name));
+        } else if (paramName === 'equalTo') {
+          q = query(q, equalTo(paramValue.index, paramValue.name));
+        } else if (paramName === 'orderBy') {
+          q = query(q, orderByChild(paramValue));
+        } else if (paramName === 'orderByKey') {
+          q = query(q, orderByKey());
+        } else if (paramName === 'orderByPriority') {
+          q = query(q, orderByPriority());
+        } else if (paramName !== 'tag') {
+          throw new Error('Unsupported query parameter: ' + paramName);
+        }
+      }
+      return q;
+    };
+
+    const testListener = function (spec) {
+      // Hack: have the callback return the spec for the listen that triggered it.
+      return function myListener() {
+        currentSpec = spec;
+      };
+    };
+
+    const checkValidProperties = function (obj, expectedProperties) {
+      for (const prop in obj) {
+        if (expectedProperties.indexOf(prop) < 0) {
+          throw new Error('Unexpected property found in spec: "' + prop + '"');
+        }
+      }
+    };
+    let currentSpec;
+    console.log('Running ' + testSpec.name);
+    let currentWriteId = 0;
+    const registrations = {};
+
+    for (let i = 0; i < testSpec.steps.length; ++i) {
+      const spec = unsafeClone(testSpec.steps[i]);
```

Comment:
Add a TODO to not clone this and create a separate object that deals with its own concerns